src/planner.c

```diff
@@ -489,22 +493,77 @@ should_optimize_query()
 
 extern void sort_transform_optimization(PlannerInfo *root, RelOptInfo *rel);
 
+static inline bool
+should_optimize_append(const Path *path)
+{
+	RelOptInfo *rel = path->parent;
+	ListCell   *lc;
+
+	if (!guc_constraint_aware_append ||
+		constraint_exclusion == CONSTRAINT_EXCLUSION_OFF)
+		return FALSE;
+
+	/*
+	 * If there are clauses that have mutable functions, this path is ripe for
+	 * execution-time optimization
+	 */
+	foreach(lc, rel->baserestrictinfo)
+	{
+		RestrictInfo *rinfo = (RestrictInfo *) lfirst(lc);
+
+		if (contain_mutable_functions((Node *) rinfo->clause))
+			return TRUE;
+	}
+	return FALSE;
+}
+
 static void
 timescaledb_set_rel_pathlist(PlannerInfo *root,
 							 RelOptInfo *rel,
 							 Index rti,
 							 RangeTblEntry *rte)
 {
-	if (extension_is_loaded() && !IS_DUMMY_REL(rel) && should_optimize_query())
+	Hypertable *ht;
+
+	if (prev_set_rel_pathlist_hook != NULL)
+		(*prev_set_rel_pathlist_hook) (root, rel, rti, rte);
+
+	if (!extension_is_loaded() ||
+		IS_DUMMY_REL(rel) ||
+		!should_optimize_query() ||
+		!OidIsValid(rte->relid))
+		return;
+
+	sort_transform_optimization(root, rel);
+
+	if (rel->rtekind != RTE_RELATION ||
+		rte->relkind != RELKIND_RELATION ||
+	/* Do not optimize result relations (INSERT, UPDATE, DELETE) */
+		rti == (Index) root->parse->resultRelation)
+		return;
+
+	ht = hypertable_cache_get_entry(gpc.hcache, rte->relid);
```

Comment:
Not sure paying cost once and potentially exiting is higher than always iterating all path nodes for every query + hypertable lookup for every append node is better. Do you see a clear benefit moving this to `should_optimize_append()`